New Media View

EricBAndrews opened the pull request:
<!-- 
Thank you for opening a pull request! 

We assume you have read CONTRIBUTING.md. If you have not, do not be surprised if your PR is rejected for reasons listed therein.

Please note that if your PR does not address an issue that was assigned to you, you are still welcome to open it; however, it will not receive merge priority and may be rejected for scope reasons.
-->

# Checklist
- [x] I have read CONTRIBUTING.md
- [x] I have described what this PR contains
- [x] If this PR alters the UI, I have attached pictures/videos
- [x] This PR addresses one or more open issues that were assigned to me:
      - closes #1219 

# Pull Request Information

Replaces `DynamicMediaView` and `LargeImageView` with `MediaView`. `MediaView` is built on an opt-in functionality paradigm; by default it returns a plain image for the URL, and the caller may enable additional complexities as needed.

This architecture offers several advantages over the previous one:
- Image behavior is now obvious at the point where `MediaView` is invoked
- Removes the "bucket brigade" anti-pattern that was emerging in the old image view hierarchy
- Improves code consistency (e.g., previously the post used `LargeImageView`, but the image viewer used `DynamicMediaView` directly to bypass default behaviors on `LargeImageView`)
- Enables granular control over image behavior without bolt-on default behavior overrides

The new `MediaView` also has some new features:
- Aspect ratio can be vertically bounded with the `verticalAspectRatioBounds` parameter (previously this was handled by tacking `.aspectRatio` to one of the old views)
- `ContentMode` can now be specified; previously only `.fit` was allowed.

`WebsitePreviewView` has been updated to use `verticalAspectRatioBounds`, closing #1219. It uses `.fill` to always take up the full width, since `.fit` looks bad with the website preview complex.

EricBAndrews committed:
remove unnecessary parameters

EricBAndrews committed:
preliminary internalLayout approach

EricBAndrews committed:
non-optional MediaLoader url

EricBAndrews committed:
withContextMenu()

EricBAndrews committed:
basic blurring/cropping

EricBAndrews committed:
replaced LargeImageView

EricBAndrews committed:
replaced DynamicMediaView

EricBAndrews committed:
removed MediaView

EricBAndrews committed:
NewMediaView renamed to MediaView

EricBAndrews committed:
image viewer now has context menu

EricBAndrews committed:
tweak behavior with blurred animated images

EricBAndrews committed:
minor tweaks

EricBAndrews committed:
framing basically works

EricBAndrews committed:
InternalMediaView implementation

EricBAndrews committed:
tidying

EricBAndrews committed:
tidying

EricBAndrews committed:
fixed context menus, tap actions still broken

EricBAndrews committed:
fixed tap actions

EricBAndrews committed:
documentation

EricBAndrews committed:
tidy

EricBAndrews committed:
fix quicklook bug

EricBAndrews committed:
fix tap action bug

EricBAndrews committed:
Merge branch 'dev' into eric/new-media-view

Sjmarf commented on Mlem/App/Views/Shared/Images/Core/MediaView+Views.swift:
```diff
@@ -0,0 +1,183 @@
+    
+    @ViewBuilder
+    var nsfwOverlay: some View {
+        if enableNsfwBlur {
+            NsfwOverlay(blurred: $blurred)
+        }
```
There was previously a short animation when blurring/unblurring an image which doesn't seem to be working anymore

EricBAndrews replied:
Nice catch!

Fixing this I also realized the `blurValue` I was using was scaling based on underlying image size, not visual size--I've reverted that back to `.dynamicBlur`.

Sjmarf commented on Mlem/App/Views/Shared/Images/Wrappers/ThumbnailImageView.swift:
```diff
@@ -8,6 +8,7 @@
 import Foundation
 import MlemMiddleware
 import SwiftUI
+import _QuickLook_SwiftUI // https://github.com/swiftlang/swift/issues/73048
```
I believe frameworks formatted like `_QuickLook_SwiftUI` are imported automatically if you import both `QuickLook` and `SwiftUI`. Would importing `QuickLook` here instead also work?

```suggestion
import QuickLook
```

Sjmarf commented on Mlem/App/Views/Shared/WebsitePreviewView.swift:
```diff
@@ -59,24 +57,22 @@ struct WebsitePreviewView: View {
-                    }
+                MediaView(
+                    url: thumbnailUrl,
+                    verticalAspectRatioBounds: .init(width: 4, height: 5),
```
Imo we should use a shorter aspect ratio here - the website preview as a whole (including the site title) ends up taller than 4:5 which looks too tall to me. Maybe 1:1 would be better?

EricBAndrews replied:
Works for me 👍

Sjmarf commented on Mlem/App/Views/Shared/Images/Core/MediaView.swift:
```diff
@@ -2,58 +2,142 @@
+            .overlay(errorOverlay)
+            .clipShape(.rect(cornerRadius: cornerRadius))
+            .withContextMenu(menuContent: contextMenuContent, isEnabled: enableContextMenu)
+            .gesture(TapGesture().onEnded(tapActions), isEnabled: onTapActions != nil)
```
This causes the image viewer to not open when no additional `onTapActions` are specified (e.g. in inline markdown images).

```suggestion
            .gesture(TapGesture().onEnded(tapActions), isEnabled: (onTapActions != nil) || enableImageViewer)
```

EricBAndrews replied:
Nice catch!

Sjmarf reviewed (changes requested):
Nice, this is super clean 😄 Found a few bugs

Sjmarf commented on Mlem/App/Utility/Extensions/MarkdownConfiguration+Extensions.swift:
```diff
@@ -68,9 +68,12 @@ extension MarkdownConfiguration {
-            .frame(maxWidth: .infinity)
+        MediaView(
+            url: inlineImage.url,
+            verticalAspectRatioBounds: .init(width: 4, height: 5),
```
```suggestion
            verticalAspectRatioBounds: .init(width: 4, height: 5),
            cornerRadius: Constants.main.mediumItemCornerRadius,
```

EricBAndrews committed:
blur animation

EricBAndrews committed:
quicklook import

EricBAndrews committed:
website complex aspect ratio 1:1

EricBAndrews committed:
Update Mlem/App/Views/Shared/Images/Core/MediaView.swift

Co-authored-by: Sjmarf <78750526+Sjmarf@users.noreply.github.com>

EricBAndrews committed:
Update Mlem/App/Utility/Extensions/MarkdownConfiguration+Extensions.swift

Co-authored-by: Sjmarf <78750526+Sjmarf@users.noreply.github.com>

EricBAndrews committed:
fix material and play button not cooperating

EricBAndrews commented:
Implementation note:

I've updated the behavior of blurred animated content so that unblurring automatically plays. There are two reasons for this:
- Smoother UX (one tap instead of two in the common case)
- The material of the `PlayButton` and the blur animation create some nasty artifacts together. I've added some code to `AnimationControlLayer` to fix this on the blur case by hiding the play button immediately when blurring so that it isn't present when the blur animation happens. This doesn't work in the unblur case, however, because the play button would need to be shown after the unblur animation completes and SwiftUI's timing code is not that precise; the play button either appears slightly before, causing artifacts, or slightly after, which looks bad.

EricBAndrews committed:
progressview tweaks

EricBAndrews committed:
more bugfix

EricBAndrews committed:
Merge branch 'dev' into eric/new-media-view

Sjmarf reviewed (approved):
Nice, LGTM 🔥
